Re-enable low memory dropout

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103330


On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @aakhundov

eellison committed:
Re-enable low memory dropout

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Re-enable low memory dropout"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Re-enable low memory dropout"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Re-enable low memory dropout"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ngimel commented on torch/_inductor/fx_passes/replace_random.py:
```diff
@@ -123,3 +131,62 @@ def replacement(size):
+    scale = float(1.0 / (1.0 - dropout_p))
+
+    def get_bool_mask():
+        return inductor_prims.random(x.size(), seed, "rand") > dropout_p
```
how would this guarantee that seed is the same between forward and backward, and different between different dropouts/different iterations?

eellison replied:
Each trace generates a new seed, and we forbid the seed generation from being recomptued - `random_ops += [prims.inductor_seeds, prims.inductor_seed]`.

eellison committed:
Update on "[WIP] Re-enable low memory dropout"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

anijain2305 commented on torch/_functorch/partitioners.py:
```diff
@@ -455,6 +457,12 @@ def classify_nodes(joint_module):
+        # this takes in a seed, so it's deterministic, and we can recompute it
+        default_recomputable_ops += [prims.inductor_random]
+
+        random_ops += [prims.inductor_seeds, prims.inductor_seed]
```
We should probably recompute seeds? This might change rng behavior.

eellison replied:
This does prevent recomputation of seeds

anijain2305 replied:
Ah, sorry, misssed it.

eellison committed:
Update on "[WIP] Re-enable low memory dropout"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Re-enable low memory dropout"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot rebase

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

malfet commented:
@pytorchbot revert -m "Lots of tests failed with 'prims' object has no attribute 'inductor_random'" -c landrace

malfet commented:
See https://hud.pytorch.org/pytorch/pytorch/commit/2d14395f176b38b8416c2713d285e5ae55695a5f

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Re-enable low memory dropout"



On attention_is_all_you_need_pytorch:

Perf: 1.526x -> 1.544x      
Memory: 1.00 -> 1.05x

Fix for https://github.com/pytorch/pytorch/issues/102319, although I'm not sure all the perf is recovered.
 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge -f "unrelated /flakey failures"

davidberard98 commented:
@eellison This caused 117s -> 149s compilation latency on dynamic huggingface [link](https://hud.pytorch.org/benchmark/compilers?startTime=Fri%2C%2023%20Jun%202023%2022%3A35%3A39%20GMT&stopTime=Fri%2C%2030%20Jun%202023%2022%3A35%3A39%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=main&lCommit=7bc181d374a54db4b1556d263be965f6a848eecb&rBranch=main&rCommit=d4e51511a0e471072359800b03e1b63f2ef2b3f6), e.g. 110s -> 386s T5Small

eellison commented:
Wow, that is a large regression for T5Small. I'll revert until we can figure that out further.

eellison commented:
@pytorchbot revert

eellison commented:
@pytorchbot revert -m "large compilation time regression"

davidberard98 commented:
@pytorchbot revert -m "large compilation time regression" -c nosignal

Chillee commented:
I'd guess it probably has something to do with preventing each dropout kernel from reusing the same kernel?
